consul/snapshot_endpoint.go

```diff
@@ -0,0 +1,203 @@
+		return nil, permissionDeniedErr
+	}
+
+	// Dispatch the operation.
+	switch args.Op {
+	case structs.SnapshotSave:
+		if !args.AllowStale {
+			if err := s.consistentRead(); err != nil {
+				return nil, err
+			}
+		}
+
+		// Set the metadata here before we do anything; this should always be
+		// pessimistic if we get more data while the snapshot is being taken.
+		s.setQueryMeta(&reply.QueryMeta)
+
+		// Take the snapshot and capture the index.
+		snap, err := snapshot.New(s.logger, s.raft)
+		reply.Index = snap.Index()
+		return snap, err
+
+	case structs.SnapshotRestore:
+		if args.AllowStale {
+			return nil, fmt.Errorf("stale not allowed for restore")
+		}
+		if err := snapshot.Restore(s.logger, in, s.raft); err != nil {
+			return nil, err
+		}
+		return ioutil.NopCloser(bytes.NewReader([]byte(""))), nil
+
+	default:
+		return nil, fmt.Errorf("unrecognized snapshot op %q", args.Op)
+	}
+}
+
+// handleSnapshotRequest reads the request from the conn and dispatches it. This
+// will be called from a goroutine after an incoming stream is determined to be
+// a snapshot request.
+func (s *Server) handleSnapshotRequest(conn net.Conn) error {
+	var args structs.SnapshotRequest
+	dec := codec.NewDecoder(conn, &codec.MsgpackHandle{})
+	if err := dec.Decode(&args); err != nil {
+		return fmt.Errorf("failed to decode request: %v", err)
+	}
+
+	var reply structs.SnapshotResponse
+	snap, err := s.dispatchSnapshotRequest(&args, conn, &reply)
+	if err != nil {
+		reply.Error = err.Error()
+		goto RESPOND
+	}
+	defer func() {
+		if err := snap.Close(); err != nil {
+			s.logger.Printf("[ERR] consul: Failed to close snapshot: %v", err)
+		}
+	}()
+
+RESPOND:
+	enc := codec.NewEncoder(conn, &codec.MsgpackHandle{})
+	if err := enc.Encode(&reply); err != nil {
+		return fmt.Errorf("failed to encode response: %v", err)
+	}
+	if snap != nil {
+		if _, err := io.Copy(conn, snap); err != nil {
+			return fmt.Errorf("failed to stream snapshot: %v", err)
+		}
+	}
+
+	return nil
+}
+
+// SnapshotRPC is a streaming client function for performing a snapshot RPC
+// request to a remote server. It will create a fresh connection for each
+// request, send the request header, and then stream in any data from the
+// reader (for a restore). It will then parse the received response header, and
+// if there's no error will return an io.ReadCloser (that you must close) with
+// the streaming output (for a snapshot). If the reply contains an error, this
+// will always return an error as well, so you don't need to check the error
+// inside the filled-in reply.
+func SnapshotRPC(pool *ConnPool, dc string, addr net.Addr,
+	args *structs.SnapshotRequest, in io.Reader, reply *structs.SnapshotResponse) (io.ReadCloser, error) {
+
+	conn, hc, err := pool.Dial(dc, addr)
+	if err != nil {
+		return nil, err
+	}
+
+	// keep will disarm the defer on success if we are returning the caller
+	// our connection to stream the output.
+	var keep bool
```

Comment:
Given the connection switches into this snapshot mode, is there a point of keeping it? Why not just always close given the infrequency of the operation?